### DISCUSSION TITLE
Targeting peers in fabric-gateway-go sdk

### DISCUSSION THREAD

**Discussion by tittuvarghese** [hyperledger/fabric-gateway#570] [Q&A]
Is there any way to select / target peers using fabric-gateway-go sdk. Following are the features / requirement which I'm trying to experiment.

1. Discovering peers. 
2. Targeting some of the peers discovered for transaction proposals. 
3. Listen for chain-code events from some of the peers.

This is primarily to solve problems of application trying to get endorsement from a peer located in another data-center / region or to have local peer preferences.

@bestbeforetoday I feel like you might be able to help in this.

**Comment by bestbeforetoday** [comment]
The client does not manage these concerns. Instead, it is handled by the Gateway service within the peer.

1. Gateway service uses service discovery information directly from the peer so it knows all network nodes. The client only needs to know the endpoint address of a Gateway service.
2. The client can specify specific orgs to target for endorsement, but this is only there for cases where the Gateway service cannot safely determine the required endorsing orgs without risking sending private data to an org that should not receive it. In general the Gateway service will determine the required endorsing orgs and prefer the highest block height peers from those orgs to ensure the most up-to-date ledger state and give the best change of successful transaction validation. For performance, it also prefers the local peer when obtaining endorsement from its own org.
3. Using the Fabric Gateway client API, chaincode events are received directly from the Gateway peer ledger. All peers for a given channel share the same ledger so there should not be any need to select specific peers.

There is a detailed description of the Gateway endorsement process in the [Fabric documentation](https://hyperledger-fabric.readthedocs.io/en/latest/gateway.html#how-the-gateway-endorses-your-transaction-proposal). Any peer can (and by default does) run the Gateway service, so the client can select an appropriate org/peer Gateway service to connect to. In a production deployment you probably should consider a proxy or ingress controller as the Gateway endpoint so clients don't need to know about specific peers or manage fail-over between them, as described in this [workshop material](https://github.com/hyperledger/fabric-samples/blob/eb787b7afa0a94f0c7a8e29b3246d62642decb0d/full-stack-asset-transfer-guide/docs/ApplicationDev/01-FabricGateway.md#production-deployment-of-fabric-gateway).
